documents/readthedoc/docs/source/Solutions/ehsm-kms/ehsm.md

```diff
@@ -0,0 +1,88 @@
+# eHSM (SGX Enclave Based Hardware Security Module)
+
+Cloud KMS (Key Management System) is a hardware-software combined system that provides customers with capabilities to create and manage cryptographic keys and control their use for their cloud services.
+
+Typically, the KMS (Key Management Service) is backed with dedicated HSM (Hardware Security Module). However, the existing hardware HSM solution is quite expensive and complex to extend on business demand. Although there are some commercial cloud HSM solutions to alleviate the problem of the scalability and availability, they still rely on the dedicated external hardware devices, and also currently there's no opensource product with low cost and scalable secure HSM solution deployed in cloud KMS use scenarios.
+
+This project is named as eHSM-KMS, which provides a product-quality reference to help the customers setup a customized KMS (Key Management System) with the capabilities to create and manage cryptographic keys in their cloud environments.
+
+It's based on Intel SGX (Software Guard eXtensions) technology that provides the near-equivalent hardware protection level of cryptographic functionalities including key generation, management inside the SGX enclave.
+
+The Customers could deploy this total solution on their own private cloud or deploy it on a public cloud like Alibaba Cloud ECS SGX instances in which we have already verified.
+
+---
+
+## Overview of the eHSM-KMS
+This below diagram depicts the high-level architecture of the eHSM-KMS,
+
+![arch-of-ehsm-kms](./img/arch-of-ehsm-kms.PNG)
+
+- eHSM-Core Enclave
+    - The main functional component that provides cryptographic operations based on SGX SDK Crypto APIs (openssl/Intel ipp libaray).
+    - The plaintext of keys are used/processed only inside this Enclave.
+
+- The eHSM-KMS manager
+    - A webservice hosted with the nodejs framework, which will send requests to eHSM-core enclave for cryptographic operations through the ffi native interfaces.
+    - Provide public cryptographic APIs via [RESTFUL interfaces](https://github.com/intel/ehsm/blob/bad3ba0525fa1c03f1cab55c48288df26df877af/docs/API_Reference.md) to the users. key generation, key importing/export, encryption/decryption, sign/verify, etc.
```

Comment:
Could you please replace the link with the readthedoc in CCzoo?

Reply:
ehsm acts as a submodule in cczoo. It indicates ehsm under cczoo is a link direct to github.com/intel/ehsm.
So that, there is no link in cczoo.

Reply:
@RodgerZhu  CCzoo should provide a whole solution. By the solution, users can be clear about how to use it in CCzoo but not jump to other links, right? What's your opinion?

Reply:
> @RodgerZhu CCzoo should provide a whole solution. By the solution, users can be clear about how to use it in CCzoo but not jump to other links, right? What's your opinion?

Yes, I agree with you.  Actually, we have a discussion on it this morning. For "build instruction" part, eHSM will put this part directly when they have a stable version, as they now will make changes in it.
For "RESTFUL interfaces", if this part is stable and ready, suggest to put this doc directly in CCZoo.

Thank You!

Reply:
Thanks. Got your points. Restful API doc is not stable, either. 
More crypto APIs will be added and the interfaces may also be redesigned.
I do believe by end of Q2, this part will be more stable. 
What do you think?  Thanks.

Reply:
ok. Got it. Let's do it in Q2. Thanks